scripts/samba-vuln-cve-2017-7494.nse

```diff
@@ -0,0 +1,531 @@
+          status, err = smb.smb_send(smbstate, smb_header, smb_params, '', overrides)
+          stdnse.debug1('Querying msrpc - status: %s , curr_path %s , result: %s',
+          status, curr_path, result)
+
+          -- parse the results from the smbstate
+          result, smb_header = smb.smb_read(smbstate)
+          _, smb_cmd, err = string.unpack("<c4 B I4", smb_header)
+          stdnse.debug1("result %s, curr_path %s, error %s", result, curr_path, err)
+
+          -- on payload execution, result will be false
+          if not result then
+            samba_cve.state = vulns.STATE.VULN
+            table.insert(samba_cve.check_results,
+              "Exploitation of CVE-2017-7494 succeeded!")
+            return
+          end
+        end
+      end
+    end
+  end
+  if samba_cve.state ~= vulns.STATE.VULN then
+    table.insert(samba_cve.check_results,
+      'Unable to find uploaded file, system may or may not be patched')
+  end
+end
+
+action = function(host,port)
+  local port = nmap.get_port_state(host,{number=smb.get_port(host),protocol='tcp'})
+
+  local version = port.version.version
+
+  -- retrieve version of samba using smb.get_os
+  if not version then
+    local status, result = smb.get_os(host)
+
+    if(status == false) then
+      return stdnse.format_output(false, result)
+    end
+
+    -- result.lanmanager contains OS version information
+    -- string returned by result.lanmanager looks like Samba 4.3.9-Ubuntu
+    -- we only want 4.3.9-Ubuntu
+    if string.match(result.lanmanager,"^Samba ") then
+      version = string.match(result.lanmanager,"^Samba (.*)")
+    else
+      return stdnse.format_output(false,
+        "Either versioning failed or samba does not exist on the port!")
+    end
+  end
+
+  stdnse.debug("Port Version: %s", port.version.version)
+
+  local result, stats
+  local response = {}
+
+  local samba_cve  = {
+    title = "SAMBA Remote Code Execution from Writable Share",
+    IDS = {CVE = 'CVE-2017-7494'},
+    risk_factor = "HIGH",
+    scores = {
+      CVSSv3 = "7.5 (HIGH) (CVSS:3.0/AV:N/AC:H/PR:L/UI:N/S:U/C:H/I:H/A:H)"
+    },
+    description = [[
+All versions of Samba from 3.5.0 onwards are vulnerable to a remote
+code execution vulnerability, allowing a malicious client to upload a
+shared library to a writable share, and then cause the server to load
+and execute it.
+]],
+    references = {
+      'https://www.samba.org/samba/security/CVE-2017-7494.html',
+    },
+    dates = {
+      disclosure = {year = '2017', month = '05', day = '24'},
+    },
+    check_results = {},
+    extra_info = {}
+  }
+
+  local report = vulns.Report:new(SCRIPT_NAME, host, port)
+  samba_cve.state = vulns.STATE.NOT_VULN
+
+  -- determine if version is vulnerable
+  determine_vuln_version(version, samba_cve)
+
+  -- only do other checks if the version is vulnerable
+  if samba_cve.state == vulns.STATE.LIKELY_VULN then
```

Comment:
How confident are we about the vulnerable versions? Are the same versions vulnerable to all distros? What if an administrator patches his system to return a different version number than the real one so his system won't get detected? (Yes, that's a corner case)

How about we add an argument `samba-vuln-cve-2017-7494.ignoreversion` that if set ignores the first check of the script?